Source/ACE.Server/Entity/Proficiency.cs

```diff
@@ -3,11 +3,14 @@
 using ACE.Entity.Enum;
 using ACE.Server.WorldObjects;
 using ACE.Server.WorldObjects.Entity;
+using log4net;
 
 namespace ACE.Server.Entity
 {
     public class Proficiency
     {
+        private static readonly ILog log = LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
+
         public static TimeSpan FullTime = TimeSpan.FromMinutes(15);
 
         public static void OnSuccessUse(Player player, CreatureSkill skill, uint difficulty)
```

Comment:
The thing about the uint version, is that it's still the main overload:

uint - 17 references (everything except healing kits and unlockables, currently)
int - 2 references (healing kits and unlockables)

The 'difficulty' parameter comes from a few different types of sources:

- CreatureSkill.Current, the main source. This is a very core field to the system, defined as uint, and has 73 references. This was the main source of the bug
- EffectiveAttack and EffectiveDefense skills - these are for physical attacks (melee/missile), and originate from DamageEvent. They are calculated by taking .Current skills, and factoring in multpliers
- Spell Power / PowerMod - this is the difficulty of casting a spell, and is stored in the client dat 
- Healing kits and unlockables, which are currently the only 2 outliers representing their skill #s as ints

The check for 'is difficulty < 0' in Proficiency is really just a simple last line of defense for the 2 int cases. Ideally, these boundary checks should happen much earlier in the system, at the point where they originate - these skill/difficulty values should never be < 0 anywhere in the system.

I started to add the min. 0 to healing kits, and unlockables looked like they should be fairly straightforward as well. However, to keep the current PR focused and fixing the core bug, while changing a minimum amount of the rest of the code, this check was kept to CreatureSkill.Current (by far the main case)

For the first iteration of this PR, there was only 1 Proficiency.OnSuccessUse function w/ no overload, and the difficulty parameter was changed to int. I then started to go through the entire system, everywhere this function was called, and started adjusting the calls so that it uses int instead of uint. That is when I learned all of the different types of calls to this function - changing everything to int would have used reversed the calling profiles from 2 casts / 17 native uint types, to 17 casts / 2 native int types.

I actually then started to investigate where all of these uint types came from, and if it would be appropriate for them to be represented as int throughout the system instead. And i see that by far, the source of most of this data is from CreatureSkill.Current (which is when the source of the bug was discovered)

So the current calling profile / data types:
- changing Proficiency.OnSuccessUse would go from 2 int->uint casts, and 17 native uints, to the reverse: 2 native ints, and 17 uint->int casts
- changing CreatureSkill.Current to int instead of uint would require examination of 78 functions which use it, and possibly adjusting data types in those functions as well

So I gravitated toward fixing the core of the bug systemically throughout the system, at the place where it originates (CreatureSkill.Current, and later adding similar "not less than 0" checks to healing kits / unlockables), while keeping a minimal footprint of code changes throughout the rest of the system

Kinda long winded, but just wanted to detail the current thought process..